src/peergos/shared/user/UserContext.java

```diff
@@ -1650,13 +1650,22 @@ private static void ensureAllowedToShare(FileWrapper file, String ourname, boole
     }
 
     @JsMethod
-    public CompletableFuture<List<FileWrapper>> getFiles(List<SharedItem> pointers) {
-        return Futures.combineAllInOrder(pointers.stream()
-                .map(s -> network.getFile(s.cap, s.owner))
-                .collect(Collectors.toList()))
-                .thenApply(res -> res.stream()
-                        .flatMap(Optional::stream)
-                        .collect(Collectors.toList()));
+    public CompletableFuture<List<Pair<SharedItem,FileWrapper>>> getFiles(List<SharedItem> pointers) {
+        List<CompletableFuture<Optional<?>>> futures = pointers.stream().map(s -> network.getFile(s)).collect(Collectors.toList());
+        CompletableFuture<List<Pair<SharedItem, FileWrapper>>> identity = CompletableFuture.completedFuture(Collections.emptyList());
+        return futures.stream().reduce(identity,
```

Comment:
we need to use combineAllInOrder to preserver order.

Reply:
you can change the original code to call.thenapply in the map to make a pair

Reply:
This IS the code from combineAllInOrder. I moved it inline because i couldn't add the Pair<>()

Reply:
This very different from the existing code. E.g. the existing doesn't explicitly allocate a list.